www/docs/client/mocking.md

```diff
@@ -0,0 +1,234 @@
+---
+id: Mocking
+title: Mocking your server
+sidebar_label: Mocking your server
+slug: /mocking
+---
+
+Mocking tRPC gives you the ability to isolate your frontend components into testing or presentation environments without having to alter how they behave.
+
+By intercepting our tRPC calls at the [Service Worker level](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) and returning responses as the component expects, mocking can help you to:
+
+- Write performant, easy to execute tests that don't rely on changing the underlying component in any way
+- Present componenets in frontend workshops like Storybook exactly as they would appear and behave in your main app
+
+The examples below use [Jest](https://facebook.github.io/jest/docs/en/tutorial-react.html), (React Testing Library)[https://github.com/testing-library/react-testing-library], and [Mock Service Worker](https://www.npmjs.com/package/msw-trpc) (MSW), but the concepts apply to any testing framework.
+
+## Creating a TRPCMockedProvider
+
+Every test for a React component that uses your `tRPC client` must make it available in React's context.
+
+- In React, you achieve this by wrapping your component tree with the `trpc.Provider` and `QueryClientProvider` components.
+- In Next.js, this step is abstracted in `_app.ts` by using `trpc.withTRPC(MyApp)`.
+
+In tests however, this context is not avaliable by default. Instead, you will need to expose this context by creating a `TRPCMockedProvider`.
+
+### 1. Initialize a new React hook
+
+Create a new file for your `MockedProvider` and import your `AppRouter` type from where your standard tRPC router is defined. This will ensure that your eventual mocking is typed to the standard signature of your entire API.
+
+```ts title='trpcMockedProvider.ts'
+import type { AppRouter } from "../path/to/server/trpc";
+```
+
+Next, create a set of React hooks from your `AppRouter` type signature with `createTRPCReact`, as all the tests and isolated environments we want to render our components in will be in a React environment.
+
+```ts title='trpcMockedProvider.ts'
+const mockedtRPC = createTRPCReact<AppRouter>({
+  unstable_overrides: {
+    useMutation: {
+      async onSuccess(opts) {
+        await opts.originalFn();
+        await opts.queryClient.invalidateQueries();
+      },
+    },
+  },
+});
+```
+
+### 2. Create your mocked tRPC provider
+
+Now we just need to create a wrapper that exposes the client. This process is similar to [the React setup](/reactjs/introduction.mdx) however we don't need to utilize the `useState` hook because our tests are not using SSR.
+
+```ts title='trpcMockedProvider.ts'
+// @ts-expect-error tRPC utilizes global fetch
+global.fetch = fetch;
+
+const mockedtRPCClient = mockedtRPC.createClient({
+  links: [
+    loggerLink({ enabled: () => false }),
+    httpBatchLink({ url: "http://localhost:3000/api/trpc" }),
```

Comment:
What it means is that when you define request handlers with `msw-trpc`, it is only listening for the specific HTTP RPC endpoint for the intended endpoint you specify from the router. When you batch things, it doesn't create any errors in this instance, however, a request handler won't exist in `msw`'s "server" to intercept the resulting HTTP call.

In this instance, you'll see a bright yellow warning in your terminal (assuming you are executing a test) letting you know that the component sent a request that wasn't intercepted. Honestly, I've struggled to get a `mockedTRPCClient` running without specifying a batch link. Would appreciate your insight if I'm doing something goofy.

Reply:
It seems to me like it will be fairly easy to trigger this issue, for example in a component that fires two different queries when it mounts. If at all possible we should try to get this working.